File: src/minilake/mcp/tools/clusters.py
"""MCP tools for Clusters (/api/2.1/clusters).

Clusters are a state machine only — there is no Spark compute behind them. Real Spark
execution happens through Jobs (see run_python_script). These tools exist so client code
that creates and polls clusters can be exercised.
"""

from typing import Any, Optional

from minilake.mcp.client import MinilakeClient

_PREFIX = "/api/2.1/clusters"

_NO_COMPUTE_NOTE = (
    "Note: minilake clusters are a state machine with no Spark compute behind them. "
    "To actually execute Spark code, use run_python_script."
)


def register(mcp: Any, client: MinilakeClient) -> None:
    @mcp.tool()
    async def create_cluster(
        cluster_name: str,
        spark_version: str = "13.3.x-scala2.12",
        node_type_id: str = "Standard_DS3_v2",
        num_workers: Optional[int] = 1,
    ) -> dict[str, Any]:
        """Create a cluster. Transitions PENDING -> RUNNING after a short real delay.

        No Spark compute is attached — this only drives the lifecycle state machine.
        """
        body = await client.post(
            f"{_PREFIX}/create",
            json={
                "cluster_name": cluster_name,
                "spark_version": spark_version,
                "node_type_id": node_type_id,
                "num_workers": num_workers,
            },
        )
        return {**body, "note": _NO_COMPUTE_NOTE}

    @mcp.tool()
    async def list_clusters() -> dict[str, Any]:
        """List all clusters and their states."""
        return await client.get(f"{_PREFIX}/list")

    @mcp.tool()
    async def get_cluster(cluster_id: str) -> dict[str, Any]:
        """Get one cluster's details and current state."""
        return await client.get(f"{_PREFIX}/get", params={"cluster_id": cluster_id})

    @mcp.tool()
    async def start_cluster(cluster_id: str) -> dict[str, Any]:
        """Start a terminated cluster (PENDING -> RUNNING)."""
        return await client.post(f"{_PREFIX}/start", json={"cluster_id": cluster_id})

    @mcp.tool()
    async def terminate_cluster(cluster_id: str) -> dict[str, Any]:
        """Terminate a cluster (TERMINATING -> TERMINATED). The record is kept."""
        return await client.post(f"{_PREFIX}/delete", json={"cluster_id": cluster_id})

File: ui/src/lib/api/clusters.ts
import { apiFetch } from "./client";
import type { Cluster, SparkVersion } from "./types";

const CLUSTERS = "/api/2.1/clusters";

export async function listClusters(): Promise<Cluster[]> {
  const data = await apiFetch<{ clusters?: Cluster[] }>(`${CLUSTERS}/list`);
  return data.clusters ?? [];
}

export async function listSparkVersions(): Promise<SparkVersion[]> {
  const data = await apiFetch<{ versions?: SparkVersion[] }>(`${CLUSTERS}/spark-versions`);
  return data.versions ?? [];
}

export function createCluster(params: {
  cluster_name: string;
  spark_version: string;
  num_workers?: number;
}): Promise<{ cluster_id: string }> {
  return apiFetch(`${CLUSTERS}/create`, { method: "POST", body: params });
}

export function startCluster(clusterId: string): Promise<unknown> {
  return apiFetch(`${CLUSTERS}/start`, { method: "POST", body: { cluster_id: clusterId } });
}

export function restartCluster(clusterId: string): Promise<unknown> {
  return apiFetch(`${CLUSTERS}/restart`, { method: "POST", body: { cluster_id: clusterId } });
}

/** Terminate. Real Databricks keeps the record for history — so does minilake. */
export function terminateCluster(clusterId: string): Promise<unknown> {
  return apiFetch(`${CLUSTERS}/delete`, { method: "POST", body: { cluster_id: clusterId } });
}

/** The only call that actually removes a terminated cluster from the list. */
export function permanentDeleteCluster(clusterId: string): Promise<unknown> {
  return apiFetch(`${CLUSTERS}/permanent-delete`, { method: "POST", body: { cluster_id: clusterId } });
}
